100 hello world

rodolakis opened the pull request:
- close #100

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "<details>\n",
+    "<summary>Not familiar with object-oriented programming glossary: class, object, method and instances?</summary>\n",
+    "\n",
+    "- A **class** is like a blueprint or a template that defines the characteristics and behaviors of a particular type of object. For example, we can define a `Dog` class that includes attributes such as breed, age, and name, as well as behaviors such as barking and wagging its tail.\n",
```
Good to cite some online references where user could learn more.  These two came up first in my search:

- https://docs.python.org/3/tutorial/classes.html
- https://www.w3schools.com/python/python_classes.asp

Link to something that matches the context of this notebook.

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "\n",
+    "- A **class** is like a blueprint or a template that defines the characteristics and behaviors of a particular type of object. For example, we can define a `Dog` class that includes attributes such as breed, age, and name, as well as behaviors such as barking and wagging its tail.\n",
+    "\n",
+    "- An **object** is an **instance** of a class. So, if we have a `Dog` class, we can create objects of that class, such as \"Fido\" and \"Buddy\". Each object of the `Dog` class will have its own set of **attributes**, such as \"Fido\" being a Golden Retriever and \"Buddy\" being a Chihuahua, and **methods**, such as \"Fido\" barking loudly and \"Buddy\" wagging its tail.\n",
```
Ok, now I have to see this as a notebook.  Here's my shortcut to the version in this commit: https://github.com/BCDA-APS/bluesky_training/blob/05ede6ce096b2b1ede93fabbd7252b5a08e20d61/docs/source/tutor/_hello_world.ipynb

rodolakis replied:
sorry I did not get that

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+   ],
+   "source": [
+    "import pathlib, sys, os\n",
+    "my_path = str(pathlib.Path.home() / \"bluesky\")\n",
```
Keep `my_path` as a pathlib object.  Maybe someone will use it later.  Change to:

```py
my_path = pathlib.Path.home() / "bluesky"
sys.path.append(str(my_path))
```

Convert to `str` when necessary.

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "\n",
+    "- `my_path = ...` define a variable `my_path` as the \"bluesky\" directory located in the user's home directory\n",
+    "- `sys.path.append(my_path)` adds this directory to the list of directories that Python will search when you try to import a module.\n",
+    "- `cat = databroker.temp().v2` creates a temporary databroker instance and assigns it to the variable `cat`. This allows you to interact with the databroker instance and access its functionality, such as querying for data, creating runs, or accessing metadata.\n",
```
creates an instance of a databroker _temporary_ catalog (a catalog which is deleted when the python session is exited).

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "\n",
+    "- A **class** is like a blueprint or a template that defines the characteristics and behaviors of a particular type of object. For example, we can define a `Dog` class that includes attributes such as breed, age, and name, as well as behaviors such as barking and wagging its tail.\n",
+    "\n",
+    "- An **object** is an **instance** of a class. So, if we have a `Dog` class, we can create objects of that class, such as \"Fido\" and \"Buddy\". Each object of the `Dog` class will have its own set of **attributes**, such as \"Fido\" being a Golden Retriever and \"Buddy\" being a Chihuahua, and **methods**, such as \"Fido\" barking loudly and \"Buddy\" wagging its tail.\n",
```
Python has a naming convention (not rules).  This might be a place to mention them.

- PEP8 (formal document): https://peps.python.org/pep-0008/#naming-conventions
- cheat sheet: https://namingconvention.org/python/

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "- `sys.path.append(my_path)` adds this directory to the list of directories that Python will search when you try to import a module.\n",
+    "- `cat = databroker.temp().v2` creates a temporary databroker instance and assigns it to the variable `cat`. This allows you to interact with the databroker instance and access its functionality, such as querying for data, creating runs, or accessing metadata.\n",
+    "- `RE = bluesky.RunEngine()` creates a new instance of the RunEngine class and assigns it to the variable `RE`. The RunEngine is the central component of Bluesky, responsible for managing the execution of experimental plans and coordinating communication between other components, such as detectors and data acquisition devices.\n",
+    "- `RE.subscribe(cat.v1.insert)` sets up a subscription to the RunEngine instance (`RE`) that will insert data into a databroker instance (`cat.v1`) when each run is complete: the data collected during each run will automatically be inserted into the databroker instance at the end of the run. This ensures that the data is stored and available for later analysis and processing.\n",
```
change `when each run is complete` to `as data is collected`

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+    "- `cat = databroker.temp().v2` creates a temporary databroker instance and assigns it to the variable `cat`. This allows you to interact with the databroker instance and access its functionality, such as querying for data, creating runs, or accessing metadata.\n",
+    "- `RE = bluesky.RunEngine()` creates a new instance of the RunEngine class and assigns it to the variable `RE`. The RunEngine is the central component of Bluesky, responsible for managing the execution of experimental plans and coordinating communication between other components, such as detectors and data acquisition devices.\n",
+    "- `RE.subscribe(cat.v1.insert)` sets up a subscription to the RunEngine instance (`RE`) that will insert data into a databroker instance (`cat.v1`) when each run is complete: the data collected during each run will automatically be inserted into the databroker instance at the end of the run. This ensures that the data is stored and available for later analysis and processing.\n",
+    "- `RE.subscribe(BestEffortCallback())` sets up a subscription to the RunEngine instance (`RE`) that will automatically print a summary of the run data at the end of each run.\n",
```
change `print a summary of the run data at the end of each run.` to `show a chart and print a table of the run data and a summary at the end of the run.`

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can retrieve data for **the most recent run** stored in the `cat` databroker instance, and assign it to the a variable `run`. The data is available until we quit the IPython session; to access data later, see <a ref=https://bcda-aps.github.io/bluesky_training/howto/_after_measurement.html> here </a>."
```
Seems as if that notebook belongs in this repo anyway.  As a _How To Guide_.

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,698 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "run = cat[-1]\n",
```
term(s) | `databroker.catalog` version | description
--- | --- | ---
`db` | v1 | original `Broker`-style databroker catalog
header, `h` | v1 | reference to the document stream for a Bluesky run
`cat` | v2 | `Intake`-style databroker catalog (this is what we use now)
`run` | v2 | reference to a Bluesky run
`run.primary.read()` | v2 | returns an xarray object filled with the run's primary data stream

prjemian replied:
I believe it is confusing to present the v1 terms.  The only places I use them now:

- `RE.subscribe(cat.v1.insert)`
- accessing the stream of documents directly: `for key, doc in h.documents(): ...`
- documentation that needs to be updated

prjemian replied:
To get the run's metadata:  `run.metadata` returns a `dict`.  The keys are: `['start', 'stop', 'catalog_dir']`

prjemian replied:
The difference between Broker and Intake style catalogs is seen by the user in how much work they must do to access the acquired data.  

The (original) Broker-style catalog focused on the ability to replay the documents in sequence as they were acquired.  As a consequence, the user had to understand how to step through these documents to build the data arrays for their measurement.

This was too tedious for most users.  The Intake-style catalog was an upgrade which focused on _using the data_ as it was organized into named streams,.

Another difference affects the ones who install and maintain the local catalog definitions.  The location and YAML format of the catalog definition file changed between Broker and Intake catalogs.  For the APS, I shifted all catalog definitions to Intake-style about a year ago.

prjemian reviewed (changes requested):
Some changes are requested, as noted.

rodolakis commented:
Added the requested changes

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "<summary>More about retrieving data</summary>\n",
+    "\n",
+    "When working with a regular (\"not temporary\") catalog, more details about working with the data can be found here: \n",
+    "- <a ref=https://bcda-aps.github.io/bluesky_training/howto/_after_measurement.html> How to access the data later, after the measurement.</a>\n",
```
Does not make a link in the online view.  Try with markdown syntax instead:

- [How to access the data later, after the measurement.](https://bcda-aps.github.io/bluesky_training/howto/_after_measurement.html)

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "\n",
+    "When working with a regular (\"not temporary\") catalog, more details about working with the data can be found here: \n",
+    "- <a ref=https://bcda-aps.github.io/bluesky_training/howto/_after_measurement.html> How to access the data later, after the measurement.</a>\n",
+    "- <a ref=https://bcda-aps.github.io/apstools/latest/examples/ho_search_databroker.html> How to find Bluesky data.</a>\n",
```
same as previous line

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "outputs": [],
+   "source": [
+    "run = cat[-1]\n",
+    "header=run.metadata\n",
```
To avoid confusion with the databroker v1 `header` term, let's use `metadata = run.metadata`

This will be consistent with the messaging that Bluesky makes  _metadata_ integral to the process.  Search relies on finding and matching _metadata_.

prjemian reviewed (changes requested):
A couple broken links and one change in names, as commented.

prjemian commented:
Here's the [web view](https://bcda-aps.github.io/bluesky_training/tutor/_hello_world.html) of _Hello, World!_.  BTW, this matches with my local Sphinx build.

prjemian commented:
![Clipboard01](https://user-images.githubusercontent.com/2279984/234664521-297c1e3d-8064-42da-a61b-dea0dc7ab91e.png)

prjemian commented:
Additional advice from [mozilla](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details#examples) looks interesting but does not work with the theme we are using (or with the sphinx-rtd-theme).

prjemian commented:
The `<details>` links look as good as we can get them now.  To change display ideas, convert each of the `<details>` items to a bullet-point refence linked to a (new) FAQ number.   Create the FAQ page, per #17, and merge the FAQ page into this branch.  Then begin the conversion as described.

prjemian commented:
Just started the workflow to publish this branch.  https://github.com/BCDA-APS/bluesky_training/actions/runs/4821152099/jobs/8586628704

rodolakis commented:
Moved details from the intro to FAQ and added _start_bluesky_session.md to tutorial with relevant link in Hello World and FAQ. I expended the details to retrieve data (I think it works well that way) but I left the one for the line by line explanation, not sure it belongs to the FAQ. If you agree with that then it should be ready to merge. If not I am open to your suggestion for the line by line thing!

prjemian commented on docs/source/tutor/_hello_world.ipynb:
```diff
@@ -222,12 +161,13 @@
-    "The data from this run is available until we quite the IPython session. Let's take a\n",
-    "look at the data.\n",
+    "When working with a regular (\"not temporary\") catalog, more details about working with the data can be found here: \n",
+    "- [How to access the data later, after the measurement.](https://bcda-aps.github.io/bluesky_training/howto/_after_measurement.html)\n",
```
bullet lists need a blank line before (and after)

prjemian commented:
![Clipboard01](https://user-images.githubusercontent.com/2279984/234902872-c2522fdb-5d70-4c13-a036-0ad80384928f.png)

Last item in FAQ is marked as "1." but should be "4.".  Could just force it.

```rst
- #4. ...
```

rodolakis commented:
Fixed the typo. I saw all the questions have "1" instead of iterating. I used the template from Nexus, were all the questions start with 

`#. question 1?`

What am I missing?

prjemian commented:
> I saw all the questions have "1" instead of iterating.

Both of these show numbers in sequence 1, 2, 3, 1

GitHub rendering: https://github.com/BCDA-APS/bluesky_training/blob/506fc8422a85b111e3f5f769eb31aab723955938/docs/source/reference/_FAQ.rst

Published rendering: https://bcda-aps.github.io/bluesky_training/reference/_FAQ.html

prjemian commented:
Aha!  It's the addition of the anchor that resets the numbering: https://github.com/BCDA-APS/bluesky_training/blob/506fc8422a85b111e3f5f769eb31aab723955938/docs/source/reference/_FAQ.rst?plain=1#L34-L38

Start next line with `#4. What does the...`

Also, change `means` to `mean`

prjemian commented:
And, add a blank line before the bullet list: https://github.com/BCDA-APS/bluesky_training/blob/506fc8422a85b111e3f5f769eb31aab723955938/docs/source/reference/_FAQ.rst?plain=1#L29-L32

prjemian reviewed (approved):
LGTM, Great Job!
